cocotb/regression.py

```diff
@@ -229,63 +230,119 @@ def next_test(self):
-        ratio_time  = self._safe_divide(sim_time_ns, real_time)
-        
+        ratio_time = self._safe_divide(sim_time_ns, real_time)
+
         self.xunit.add_testcase(name=test.funcname,
                                 classname=test.module,
                                 time=repr(real_time),
                                 sim_time_ns=repr(sim_time_ns),
                                 ratio_time=repr(ratio_time))
 
+        # score test
+        result_pass, sim_failed = self._score_test(test, test._outcome)
+
+        # Save results
+        self._store_test_result(test.__module__, test.__name__, result_pass, sim_time_ns, real_time, ratio_time)
+        if not result_pass:
+            self.xunit.add_failure()
+            self.failures += 1
+
+        # Fail if required
+        if sim_failed:
+            self.tear_down()
+
+        self.execute()
+
+    def _init_test(self, test_func):
+        """
+        Initializes a test.
+
+        Records outcome if the initialization fails.
+        Records skip if the test is skipped.
+        Saves the initialized test if it sucessfully initializes.
+        """
+        test_init_outcome = cocotb.outcomes.capture(test_func, self._dut)
+
+        if isinstance(test_init_outcome, cocotb.outcomes.Error):
+            self.log.error("Failed to initialize test %s" % test_func.name, exc_info=True)
+            self.xunit.add_testcase(name=test_func.name,
+                                    classname=test_func.__module__,
+                                    time="0.0",
+                                    sim_time_ns="0.0",
+                                    ratio_time="0.0")
+            result_pass, sim_failed = self._score_test(test_func, test_init_outcome.without_frames(['capture']))
+            # Save results
+            self._store_test_result(test_func.__module__, test_func.__name__, result_pass, 0.0, 0.0, 0.0)
+            if not result_pass:
+                self.xunit.add_failure()
+                self.failures += 1
+            # Fail if required
+            if sim_failed:
+                self.tear_down()
+
+        else:
+            test = test_init_outcome.get()
+            if test.skip:
+                self.log.info("Skipping test %s" % test_func.name)
+                self.xunit.add_testcase(name=test_func.name,
+                                        classname=test.module,
+                                        time="0.0",
+                                        sim_time_ns="0.0",
+                                        ratio_time="0.0")
+                self.xunit.add_skipped()
+                self.skipped += 1
+                self._store_test_result(test.module, test_func.name, None, 0.0, 0.0, 0.0)
+            else:
+                self._queue.append(test)
+                self.ntests += 1
+
+    def _score_test(self, test, outcome):
+        """ Given a test and the test's outcome, determine if the test met expectations """
```

Comment:
Please include the fact that this function is responsible for logging the result too.